### PR TITLE
Config File Syntax: Extend Embedded Ruby Code support for Hashes and Arrays

### LINKED ISSUE #4385: Enable string interpolation for hash-type parameters
### Is your feature request related to a problem? Please describe.

The [FAQ](https://docs.fluentd.org/quickstart/faq#how-can-i-use-environment-variables-to-configure-parameters-dynamically) mentions that you can use string interpolation to configure parameters using environment variables, e.g. `some_field "#{ENV['FOO_HOME']}"`, but this doesn't seem to apply to keys or values in `hash`-type parameters.

### Describe the solution you'd like

It would be nice if string values (or keys for that matter) in `hash` parameters could also get interpolated in the way `string` parameters already do. For example let's say I want to use an HTTP output with an authorization header that includes a token passed in through the environment:

```
<match>
  @type http
  headers {"Authorization": "Bearer #{ENV['SOME_ENV_VAR']}"}
</match>
```

### Describe alternatives you've considered

For my particular example, a first-class feature to support bearer tokens in the HTTP output plugin (#3587) would likely solve my immediate problem, but also implementing more general-purpose string interpolation for hash parameters would remove the need for updates to that output plugin

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Athishpranav2003**
**Which issue(s) this PR fixes**: 
Fixes #4385 

**What this PR does / why we need it**: 
Support [Embedded Ruby Code](https://docs.fluentd.org/configuration/config-file#embedded-ruby-code) for Hashes and Arrays.

```
key ["foo","#{1 + 1}"] => key ["foo","2"]
key {"foo":"#{1 + 1}"} => key {"foo":"2"}
```

This is not backward compatible.
We can disable this feature by surrounding the entire value with single quotes.

```
key `["foo","#{1 + 1}"]` => key ["foo","#{1 + 1}"]
key `{"foo":"#{1 + 1}"}` => key {"foo":"#{1 + 1}"}
```

**Note: this feature is for JSON literals**

This feature is for literals that using `[]` or `{}`.

* We need to sort out the `literal` and `value` of the Fluentd Config Syntax.
  * Fluentd first interprets `literal` by parsing the config file.
  * `literal` is String or JSON.
  * Then, Fluentd interprets `value` by parsing the `literal`.
  * `value` has [various types](https://docs.fluentd.org/configuration/config-file#supported-data-types-for-values).
* **This feature has nothing to do with parsing the `value`.**

For example, we can specify Array/Hash `value` by using **String** `literal`
We don't necessarily need to use **JSON** `literal` to specify Array/Hash `value`.
The following formats works from previous versions, and there is no specification change at all.

```
key foo,bar
key foo:bar
key "foo,#{1 + 1}"
key "foo:#{1 + 1}"
```

**Note: support only String of JSON**

For example, this does not support Number of JSON.

```
key ["foo",2] => key ["foo",2]
key ["foo",#{1 + 1}] => ERROR
```

It is because JSON `literal` supports multiple lines and comments.
`#` not in String is considered the start of a comment line.

```
key ["foo", # comment
"bar", # comment
"boo" # comment
]
```

**Docs Changes**:
TODO

**Release Note**: 
* Config File Syntax: Extend Embedded Ruby Code support for Hashes and Arrays
  * Example: `key {"foo":"#{1 + 1}"} => key {"foo":"2"}`
  * Please note that this is not backward compatible, although we assume that this will never affect to actual existing configs.
  * In case the behavior changes unintentionally, you can disable this feature by surrounding the entire value with single quotes.
    * `key '{"foo":"#{1 + 1}"}' => key {"foo":"#{1 + 1}"}`

### CODE REVIEW

**Comment by Athishpranav2003** [comment]
@ashie please check this whenever you are free

**Review by daipom — COMMENTED** [review]
Thanks for this PR!

As I commented on the original issue, we need to consider the specifications in `LiteralParser` carefully.
I'd like to think about it some more, but at the moment it doesn't seem like a good design decision to give `types.rb` the same functionality as `LiteralParser`.

I would like to hear from other maintainers.

https://github.com/fluent/fluentd/issues/4385#issuecomment-2270749268

> We need to consider this as an issue of [Embedded Ruby Code](https://docs.fluentd.org/configuration/config-file#embedded-ruby-code). Using environment variables is just one way of it.
> 
> `Embedded Ruby Code` feature is implemented in `LiteralParser`:
> 
> https://github.com/fluent/fluentd/blob/e763c0761c44d9734b6aa374371387a2e8406522/lib/fluent/config/literal_parser.rb#L183-L199
> 
> https://github.com/fluent/fluentd/blob/e763c0761c44d9734b6aa374371387a2e8406522/lib/fluent/config/literal_parser.rb#L108-L110
> 
> Since it is for `double_quoted_string`, we need to use double quotes for the entire value to use this feature.
> 
> It may be possible to address this as a `types` problem, as in #4580, but the specifications in `LiteralParser` need to be carefully considered.

**Comment by ashie** [comment]
> but at the moment it doesn't seem like a good design decision to give `types.rb` the same functionality as `LiteralParser`.

I don't yet see the detail but I feel same smell.

**Comment by Athishpranav2003** [comment]
@daipom do you suggest that we should handle this scenario in `literal parser.rb`?

**Comment by daipom** [comment]
> @daipom do you suggest that we should handle this scenario in `literal parser.rb`?

Yes. We should start by considering whether there are points in `LiteralParser` that can be improved.

**Comment by Athishpranav2003** [comment]
Oh ok
let me look at it today and check if i can move this processing to it

**Comment by daipom** [comment]
@Athishpranav2003 Thanks so much!

**Comment by Athishpranav2003** [comment]
@daipom not sure of this latest changes i have pushed.
2 UTs are failing for the latest code changes which i have made sure i have reused the code in clean form.
Please suggest on what to do for this

**Comment by Athishpranav2003** [comment]
The 2 UTs for which my code is failing are
```ruby
  test('"#{"}"}"') { assert_text_parsed_as("}", '"#{"}"}"') }
  test('"#{#}"') { assert_parse_error('"#{#}"') }  # error in embedded ruby code
```
Seems like these corner cases where written specific to the previous Embedded Ruby code parsing (which i felt needs slight revamp)

**Comment by Athishpranav2003** [comment]
@daipom ignore the prev comments
I have corrected it and pushed
Also tested locally in my side. Please check it and let me know if there is any issue

**Comment by Athishpranav2003** [comment]
@daipom not sure why 2 workflows alone failed, could you check that part alone?

**Comment by daipom** [comment]
> @daipom not sure why 2 workflows alone failed, could you check that part alone?

No problem! They have nothing to do with this PR!
Sorry, some tests are unstable and sometimes fail.

**Comment by ashie** [comment]
Could you add unit test for this?

**Comment by Athishpranav2003** [comment]
@ashie added UTs for the same

**Comment by daipom** [comment]
@Athishpranav2003 
> @daipom ignore the prev comments I have corrected it and pushed Also tested locally in my side. Please check it and let me know if there is any issue

Thanks!
Could you please clarify the specification change?
The focus is on what to do with the specifications.

The current specification is:

* String
  * non-quote: Simple one-line string without spaces.
  * single-quote: Allow spaces or multiple lines.
  * double-quote: Allow escaped characters or embedded codes.
* JSON
  * The specification seems unclear to me, but looks like it is based on `JSON.parse()` and does not support escaped characters or embedded codes.

As far as seeing changes in the code, embedded codes will always be supported, right?
If so, I have 2 concerns:

* It will be impossible to write `#{...}` as a raw string in JSON.
* Not backward compatible.

We need to consider whether the new specifications are acceptable while taking into account these concerns.

**Comment by Athishpranav2003** [comment]
Strings stay the same - only double quoted strings support string interpolation

For Hashes and Arrays - We do string interpolation for anything that has `#{...}` in it. What i can do is maybe restrict this to double quoted strings inside hashes but not sure if its needed. Please let me know on this alone

Tho we have one benefit that the docs present for now we didn't restrict the interpolation to strings so it wont go against the specifications from the past

**Comment by daipom** [comment]
> What i can do is maybe restrict this to double quoted strings inside hashes but not sure if its needed. Please let me know on this alone

This is the difficult part.
JSON uses double quotes for strings.
So, we can't switch the enabled and disabled of embedded codes in with the type of quates...

**Comment by daipom** [comment]
> Strings stay the same - only double quoted strings support string interpolation
> 
> For Hashes and Arrays - We do string interpolation for anything that has `#{...}` in it.

OK. I think this direction is possible for us!
There are concerns(https://github.com/fluent/fluentd/pull/4580#issuecomment-2272700592):
 
>     * It will be impossible to write `#{...}` as a raw string in JSON.
>     * Not backward compatible.

but I think these points are less important for this feature.
To write `#{...}` as is, we can use single-quote to the entire string, such as `param '{"k","#{foo}"}'`.
So, there is a workaround.

We need to hear from other maintainers on this point.

**Comment by Athishpranav2003** [comment]
Fine
We can wait for other's to put their opinions

**Comment by daipom** [comment]
The focus is on what extent we care about the possibility that existing users are using settings like:

```xml
<match test>
  @type http
  endpoint foo
  headers {"foo": "foo#{bar"}
</match>
```

When `#{` exists, this change can break the existing setting.

File: lib/fluent/config/literal_parser.rb [hunk 1]
```diff
@@ -254,8 +254,15 @@ def scan_json(is_array)
               buffer << line_buffer + "\n"
               line_buffer = ""
             else
-              # '#' is a char in json string
-              line_buffer << char
+              if @ss.exist?(/\{[^}]+\}/)
```
Comment (Athishpranav2003) [inline lib/fluent/config/literal_parser.rb:257]: @daipom this check can be changed to capture the case you mentioned. So here we can make sure that `}` is not being matched with the json's ending `}`. In that way we avoid capturing the strings that dont have closing braces for interpolation

**Comment by ashie** [comment]
> > ```
> > * It will be impossible to write `#{...}` as a raw string in JSON.
> > * Not backward compatible.
> > ```
> 
> but I think these points are less important for this feature. To write `#{...}` as is, we can use single-quote to the entire string, such as `param '{"k","#{foo}"}'`. So, there is a workaround.

It seems that such workaround isn't available for Array & Hash types.
Following change breaks the test, since it's tried to parse as String type which isn't matched with declaration in `config_param`.

```diff
diff --git a/test/config/test_config_parser.rb b/test/config/test_config_parser.rb
index fe29028e..b0c18fc9 100644
--- a/test/config/test_config_parser.rb
+++ b/test/config/test_config_parser.rb
@@ -483,8 +483,8 @@ module Fluent::Config
           param_size 4k
           param_bool true
           param_time 10m
-          param_hash { "key1": "value1", "key2": 2 }
-          param_array ["value1", "value2", 100]
+          param_hash '{ "key1": "value1", "key2": 2 }'
+          param_array '["value1", "value2", 100]'
           param_regexp /pattern/
         ])
         target = AllTypes.new.configure(conf)
```

**Comment by ashie** [comment]
> I'm not sure but it seems that such workaround isn't available for Array & Hash types. For example following change breaks the test, since it's tried to parse as String type which isn't matched with declaration in `config_param`.

Sorry, please ignore my comment.
Even it's parsed as String, it will be parsed again as Array or Hash at config/types.rb.

**Comment by ashie** [comment]
> We need to hear from other maintainers on this point.

In general, we intend to keep compatibility so that users can continue to use same major version without changing settings. When we have to change behaviour in any reason,  it would be better to show announcements or warnings before making the change, and wait a few minor versions to make the change in actual.

On the other hand, using `#{...}` in fluentd's setting is expected to evaluate Ruby code in most cases and I guess using such syntax in other purpose seems very rare. If we investigate major plugins and it seems unlikely that such syntax will be used, we can made this change in next minor version (without adding any warning) I guess.

**Comment by ashie** [comment]
> > I'm not sure but it seems that such workaround isn't available for Array & Hash types. For example following change breaks the test, since it's tried to parse as String type which isn't matched with declaration in `config_param`.
> 
> Sorry, please ignore my comment. Even it's parsed as String, it will be parsed again as Array or Hash at config/types.rb.

It would be better adding a test for this.

**Comment by ashie** [comment]
> On the other hand, using `#{...}` in fluentd's setting is expected to evaluate Ruby code in most cases and I guess using such syntax in other purpose seems very rare. If we investigate major plugins and it seems unlikely that such syntax will be used, we can made this change in next minor version (without adding any warning) I guess.

BTW my impression is very positive for including this change in the next minor release (with addressing some corner cases as @daipom mentioned).
It seems it's much natural than before for users.

**Comment by Athishpranav2003** [comment]
> > > I'm not sure but it seems that such workaround isn't available for Array & Hash types. For example following change breaks the test, since it's tried to parse as String type which isn't matched with declaration in `config_param`.
> > 
> > 
> > Sorry, please ignore my comment. Even it's parsed as String, it will be parsed again as Array or Hash at config/types.rb.
> 
> It would be better adding a test for this.

A test to check whether its parsed as array/hash or string?

**Comment by ashie** [comment]
> A test to check whether its parsed as array/hash or string?

A test that checks whether a single quoted string is correctly parsed as array/hash when the declared type is array/hash.

**Comment by Athishpranav2003** [comment]
@ashie i guess current code renders both " and ' quoted hashes the same :( . This is because we dont handle the case seperately since its optional for config to enclose hash/array in quotes. Not sure which is correct way to solve this

File: lib/fluent/config/literal_parser.rb (hunk 1 above)
Comment (daipom) [reply lib/fluent/config/literal_parser.rb:257]: Hmm, we would be happy to reduce the risk simply, but isn't it difficult to narrow down the exact scope of the search for the end brace?
(Don't we need to analyze JSON structure?)

**Comment by daipom** [comment]
> @ashie i guess current code renders both " and ' quoted hashes the same :( . This is because we dont handle the case seperately since its optional for config to enclose hash/array in quotes. Not sure which is correct way to solve this

When we use `"` or `'` to an entire literal, that literal will be handled as String even if it is actually JSON(Array).
LiteralParser does not necessarily parse JSON(Array) as JSON because it will be parsed as JSON later when the type is applied.

https://github.com/fluent/fluentd/blob/5618dd0e86bf0bc47a6bed2da50e37943ac2c8e9/lib/fluent/config/types.rb#L203-L207

File: lib/fluent/config/literal_parser.rb (hunk 1 above)
Comment (Athishpranav2003) [reply lib/fluent/config/literal_parser.rb:257]: I guess only nearest thing we can do here is make sure we dont capture end braces with the interpolation braces.
TBH this problem boils down to proper parsing problem like how compilers do if we are thinking in that rigorous perspective

**Comment by Athishpranav2003** [comment]
https://github.com/fluent/fluentd/blob/master/lib/fluent/config/literal_parser.rb#L67-L78
@daipom 
So here we are giving higher precedence to Arrays and JSON than strings. So even if we specify the quotes its gonna get analysed as JSON/Arrays. 
After the analysis it `JSON.dumps` the array which once converts it back to string which is then again handled in `config/types.rb`

**Comment by daipom** [comment]
> https://github.com/fluent/fluentd/blob/master/lib/fluent/config/literal_parser.rb#L67-L78 @daipom So here we are giving higher precedence to Arrays and JSON than strings. So even if we specify the quotes its gonna get analysed as JSON/Arrays. After the analysis it `JSON.dumps` the array which once converts it back to string which is then again handled in `config/types.rb`

@Athishpranav2003 
If the first character of the literal is `[` or `{`, the literal will be parsed by `scan_json()`.
On the other hand, if the first character is not `[` or `{`, the literal will be parsed by `scan_string()`.

https://github.com/fluent/fluentd/blob/5618dd0e86bf0bc47a6bed2da50e37943ac2c8e9/lib/fluent/config/literal_parser.rb#L70-L76

So, if we use quotes to the entire JSON literal, such as `'{"foo":"bar"}'` , it will be parsed by `scan_string()`.
So, it will be a workaround to write `#{...}` as is.

**Comment by Athishpranav2003** [comment]
But isn't the code using skip to identify the starting { and [ ?
Since it's using the skip the " at the start will anyways match?

**Comment by ashie** [comment]
@Athishpranav2003 Please try this:
```diff
diff --git a/test/config/test_literal_parser.rb b/test/config/test_literal_parser.rb
index c5b9b188..275e468e 100644
--- a/test/config/test_literal_parser.rb
+++ b/test/config/test_literal_parser.rb
@@ -319,6 +319,9 @@ EOA
         assert_text_parsed_as('{"a":"foo1"}', '{"a":"foo#{worker_id}"}')
         ENV.delete('SERVERENGINE_WORKER_ID')
       }
+      test('no quote') { assert_text_parsed_as_json({'a'=>'b','c'=>'test'}, '{"a":"b","c":"#{v1}"}') }
+      test('single quote') { assert_text_parsed_as_json({'a'=>'b','c'=>'#{v1}'}, '\'{"a":"b","c":"#{v1}"}\'') }
+      test('double quote') { assert_text_parsed_as_json({'a'=>'b','c'=>'test'}, '"{\"a\":\"b\",\"c\":\"#{v1}\"}"') }
       json_hash_with_comment = <<EOH
 {
  "a": 1, # this is a
```

```console
$ bundle exec rake test TEST=test/config/test_literal_parser.rb
```

All tests will be passed.
It means

* test single quote: Embedded ruby code isn't evaludated, the string `#{v1}` is output as is.
* test double quote: Embedded ruby code is evaludated, the string `#{v1}` is expanded as `test`.

**Comment by daipom** [comment]
> But isn't the code using skip to identify the starting { and [ ? 
> Since it's using the skip the " at the start will anyways match?

I'm talking about this behavior, is this as expected?

```rb
$ irb -r strscan
irb(main):001> ss = StringScanner.new('{"foo":"bar"}')
=> #<StringScanner 0/13 @ "{\"foo...">
irb(main):002> ss.rest
=> "{\"foo\":\"bar\"}"
irb(main):003> ss.skip(/\{/)
=> 1
irb(main):004> ss = StringScanner.new('\'{"foo":"bar"}\'')
=> #<StringScanner 0/15 @ "'{\"fo...">
irb(main):005> ss.rest
=> "'{\"foo\":\"bar\"}'"
irb(main):006> ss.skip(/\{/)
=> nil
irb(main):007> ss.skip(/'/)
=> 1
```

**Comment by Athishpranav2003** [comment]
@ashie i tried your tests
Its passing
I guess this seems fine to me now

**Comment by Athishpranav2003** [comment]
> > But isn't the code using skip to identify the starting { and [ ?
> > Since it's using the skip the " at the start will anyways match?
> 
> I'm talking about this behavior, is this as expected?
> 
> ```ruby
> $ irb -r strscan
> irb(main):001> ss = StringScanner.new('{"foo":"bar"}')
> => #<StringScanner 0/13 @ "{\"foo...">
> irb(main):002> ss.rest
> => "{\"foo\":\"bar\"}"
> irb(main):003> ss.skip(/\{/)
> => 1
> irb(main):004> ss = StringScanner.new('\'{"foo":"bar"}\'')
> => #<StringScanner 0/15 @ "'{\"fo...">
> irb(main):005> ss.rest
> => "'{\"foo\":\"bar\"}'"
> irb(main):006> ss.skip(/\{/)
> => nil
> irb(main):007> ss.skip(/\'/)
> => 1
> ```

Not sure
I can check this but the problem seems to be solved now tho :0

**Comment by daipom** [comment]
Thanks! I'm checking the acutual behavior.

**Comment by daipom** [comment]
# Array behavior

Applied to only explicit arrays, except for single quoted arrays.

Config:

```xml
<source>
  @type sample
  tag test1
  sample {"k1":"v1","k2":"v2","k3":"v3"}
</source>

<source>
  @type sample
  tag test2
  sample {"k1":"v1","k2":"v2","k3":"v3"}
</source>

<source>
  @type sample
  tag test3
  sample {"k1":"v1","k2":"v2","k3":"v3"}
</source>

# Not applied to shorthand array
<filter test1>
  @type record_transformer
  remove_keys k1, "#{ENV['FOO']}"
</filter>

# Applied to array
<filter test2>
  @type record_transformer
  remove_keys ["k1", "#{ENV['FOO']}"]
</filter>

# Not applied to single quoted array
<filter test3>
  @type record_transformer
  remove_keys '["k1", "#{ENV[\'FOO\']}"]'
</filter>

<match test*>
  @type stdout
</match>
```

Parsed result:

```xml
<ROOT>
  <source>
    @type sample
    tag "test1"
    sample {"k1":"v1","k2":"v2","k3":"v3"}
  </source>
  <source>
    @type sample
    tag "test2"
    sample {"k1":"v1","k2":"v2","k3":"v3"}
  </source>
  <source>
    @type sample
    tag "test3"
    sample {"k1":"v1","k2":"v2","k3":"v3"}
  </source>
  <filter test1>
    @type record_transformer
    remove_keys k1, "#{ENV['FOO']}"
  </filter>
  <filter test2>
    @type record_transformer
    remove_keys ["k1","k2"]
  </filter>
  <filter test3>
    @type record_transformer
    remove_keys ["k1", "#{ENV['FOO']}"]
  </filter>
  <match test*>
    @type stdout
  </match>
</ROOT>
```

Output:

```
... test1: {"k2":"v2","k3":"v3"}
... test2: {"k3":"v3"}
... test3: {"k2":"v2","k3":"v3"}
```

**Comment by daipom** [comment]
# Hash behavior

Applied to only explicit hashes, except for single quoted hashes.

It should be noted that it can also be applied to non-string values of JSON.
It may feel strange because non-String values are not surrounded by double quotes.
The same thing is true for Array.
However, this behavior looks acceptable to me because non-string values of JSON would rarely be used.

Config:

```xml
# Not Applied to shorthand hash
<match test>
  @type http
  endpoint foo
  headers key:#{ENV['FOO']}
</match>

# Applied to String value of hash
<match test>
  @type http
  endpoint foo
  headers {"key":"#{ENV['FOO']}"}
</match>

# Applied to non-String value of hash
<match test>
  @type http
  endpoint foo
  headers {"key":#{ENV['FOO']}}
</match>

# Not Applied to single quoted hash
<match test>
  @type http
  endpoint foo
  headers '{"key":"#{ENV[\'FOO\']}"}'
</match>
```

Parsed result (`FOO=0`):

```xml
<ROOT>
  <match test>
    @type http
    endpoint "foo"
    headers key:#{ENV['FOO']}
  </match>
  <match test>
    @type http
    endpoint "foo"
    headers {"key":"0"}
  </match>
  <match test>
    @type http
    endpoint "foo"
    headers {"key":0}
  </match>
  <match test>
    @type http
    endpoint "foo"
    headers {"key":"#{ENV['FOO']}"}
  </match>
</ROOT>
```

Note

* `in_sample` has the original logic to parse the `sample` param, so it is unsuitable for testing this behavior.
* `worker_id` is not set when checking config on launching, so it can cause a parse error if used for a non-string value of JSON.
  * We will need some fixes in the future for this problem.

```xml
<match test>
  @type http
  endpoint foo
  headers {"key":#{worker_id}}
</match>
```

```
.../home/daipom/work/fluentd/fluentd/lib/fluent/config/basic_parser.rb:92:in `parse_error!': got incomplete JSON hash configuration at 4.conf line 7,1 (Fluent::ConfigParseError)
```

**Comment by daipom** [comment]
https://github.com/fluent/fluentd/pull/4580#issuecomment-2277207381
> It should be noted that it can also be applied to non-string values of JSON.
> It may feel strange because non-string values are not surrounded by double quotes.
> The same thing is true for Array.
> However, this behavior looks acceptable to me because non-string values of JSON would rarely be used.

Considering this point, it is inaccurate to describe this PR's specification as `string interpolation fix`.

https://github.com/fluent/fluentd/pull/4580#issuecomment-2272820976
> For Hashes and Arrays - We do string interpolation for anything that has #{...} in it.

We should describe this specification as 

* Support [Embedded Ruby Code](https://docs.fluentd.org/configuration/config-file#embedded-ruby-code) for Hashes and Arrays.
  * Now, we can evaluate the Ruby code with `#{}` in `"` quoted string or in explicit Hash/Array.
  * This can be disabled by surrounding the entire value with single quotes.

**Comment by Athishpranav2003** [comment]
@daipom i will change the subject of PR as you mentioned. 
For the worker_id should we also add that check or should we give that responsibility to users?

**Comment by daipom** [comment]
> i will change the subject of PR as you mentioned. 

Thanks!!

> For the worker_id should we also add that check or should we give that responsibility to users?

I think this PR does not need to care about the `worker_id` problem!
It does not affect the existing users and it should be a rare case.

**Comment by daipom** [comment]
@Athishpranav2003 Thanks! This looks good to me!
Let me discuss the compatibility concerns a bit more with the other maintainers.

**Comment by Athishpranav2003** [comment]
> @Athishpranav2003 Thanks! This looks good to me! Let me discuss the compatibility concerns a bit more with the other maintainers.

Peace :). 
Let me know if anything needs to be changed

File: lib/fluent/config/literal_parser.rb (hunk 1 above)
Comment (ashie) [inline lib/fluent/config/literal_parser.rb:257]:
```suggestion
              if @ss.exist?(/^\{[^}]+\}/)
```
Need to  match `{` just after `#`.

**Review by ashie — CHANGES_REQUESTED** [review]

**Comment by Athishpranav2003** [comment]
@ashie i have added the changes you suggested.

**Review by daipom — APPROVED** [review]
LGTM. Thanks!

**Review by ashie — APPROVED** [review]
LGTM

Waiting to start merge window for v1.18.0.

**Comment by daipom** [comment]
I have updated [the description](https://github.com/fluent/fluentd/pull/4580#issue-2450026010).

**Comment by daipom** [comment]
TODO document.

**Comment by cosmo0920** [comment]
Yay, it's a really good enhancement! :+1 :+1:
